From: Lionel Le Folgoc <mrpouit@ubuntu.com>
Date: Fri, 26 Jul 2024 12:18:43 +0200
Subject: retrieve translations with the X-Ubuntu-Gettext-Domain key

Forwarded: not-needed

Display translations for patched desktop files from Ubuntu whose
translations are stored in a different place ("X-Ubuntu-Gettext-Domain"
key is used to find it).
---
 libxfce4util/xfce-rc.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/libxfce4util/xfce-rc.c b/libxfce4util/xfce-rc.c
index ebb35c6..7cbe9b7 100644
--- a/libxfce4util/xfce-rc.c
+++ b/libxfce4util/xfce-rc.c
@@ -555,11 +555,30 @@ xfce_rc_read_entry (const XfceRc *rc,
                     const gchar  *fallback)
 {
   const gchar *value;
+#ifdef UBUNTU
+  gboolean     has_gettext_domain;
+  const gchar *gettext_domain;
+#endif
 
   g_return_val_if_fail (rc != NULL, fallback);
   g_return_val_if_fail (rc->read_entry != NULL, fallback);
   g_return_val_if_fail (key != NULL, fallback);
 
+#ifdef UBUNTU
+  has_gettext_domain = xfce_rc_has_entry (rc, "X-Ubuntu-Gettext-Domain");
+
+  if (has_gettext_domain == TRUE && (strcmp (key, "Name") == 0 || strcmp (key, "GenericName") == 0 || strcmp (key, "Comment") == 0))
+  {
+    gettext_domain = (*rc->read_entry) (rc, "X-Ubuntu-Gettext-Domain", FALSE);
+
+    if (gettext_domain != NULL)
+    {
+      value = dgettext (gettext_domain, xfce_rc_read_entry_untranslated (rc, key, fallback));
+      return value;
+    }
+  }
+#endif
+
   value = (*rc->read_entry) (rc, key, TRUE);
   if (value == NULL)
     value = fallback;
